### PR TITLE
'avro.java.string' support in  fast-avro

### PR DESCRIPTION
**Pull request by gaojieliu**
@radai-rosenblatt @FelixGV @volauvent

### CODE REVIEW

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java [hunk 1]
```diff
@@ -63,7 +63,7 @@
   public SchemaAssistant(JCodeModel codeModel, boolean useGenericTypes, Class defaultStringClass) {
     this.codeModel = codeModel;
     this.useGenericTypes = useGenericTypes;
-    this.exceptionsFromStringable = new HashSet<>();
+    this.exceptionsFromStringable = new TreeSet<>(Comparator.comparing(Class::getCanonicalName));
```
Comment (volauvent) [inline avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java:66]: curious to know why changing to use `TreeSet` here?

File: avro-fastserde/src/test/java/com/linkedin/avro/fastserde/FastStringableTest.java [hunk 2]
```diff
@@ -132,6 +136,30 @@ public void serializeStringableFields() throws URISyntaxException, MalformedURLE
...
+    if (whetherUseFastDeserializer) {
+      afterDecoding = readWithFastAvro(schema, schema, decoder, false);
+    } else {
+      afterDecoding = readWithSlowAvro(schema, schema, decoder, false);
+    }
+
+    if (Utils.isAvro14()){
+      Assert.assertTrue(afterDecoding.get(0) instanceof Utf8, "Utf8 is expected, but got: " + afterDecoding.get(0).getClass());
+    }  else {
+      Assert.assertTrue(afterDecoding.get(0) instanceof String, "String is expected, but got: " + afterDecoding.get(0).getClass());
```
Comment (volauvent) [inline avro-fastserde/src/test/java/com/linkedin/avro/fastserde/FastStringableTest.java:159]: I may miss some context here, so `String` would be the default format after Avro 1.4?

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java (hunk 1 above)
Comment (gaojieliu) [reply avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java:66]: I should add some comments here, and this is used to make the code gen deterministic in FastDeserializerGenerator.java:
```
for (Class<? extends Exception> classException : schemaAssistant.getExceptionsFromStringable()) {
          JCatchBlock catchBlock = tryBlock._catch(codeModel.ref(classException));
          JVar exceptionVar = catchBlock.param("e");
          catchBlock.body()._throw(JExpr._new(codeModel.ref(AvroRuntimeException.class)).arg(exceptionVar));
        }
```

File: avro-fastserde/src/test/java/com/linkedin/avro/fastserde/FastStringableTest.java (hunk 2 above)
Comment (gaojieliu) [reply avro-fastserde/src/test/java/com/linkedin/avro/fastserde/FastStringableTest.java:159]: Avro-1.4 doesn't recognize the modern hint  defined in the schema: `avro.java.string`, so the de-serialized result will  be Utf8, and those hints were introduced in avro-1.7, and Avro will convert the string field to Java String.
GenericDatumReader.java in avro-1.7:
```
protected Class findStringClass(Schema schema) {
    String name = schema.getProp(GenericData.STRING_PROP);
    if (name == null) return CharSequence.class;

    switch (GenericData.StringType.valueOf(name)) {
      case String:
        return String.class;
      default:
        return CharSequence.class;
    }
  }
```

**Pull request merged**